File: Fullstack/Full-Stack/backend-nutrify/src/controllers/chat.controller.js
import mongoose from "mongoose";

import * as chatService from "../services/chat.service.js";

const isValidString = (value) => typeof value === "string" && value.trim().length > 0;

export const sendChatMessage = async (req, res) => {
  try {
    const { message, conversationId } = req.body;
    const userId = req.user._id; // Set securely from JWT token

    if (!isValidString(message)) {
      return res.status(400).json({
        success: false,
        message: "Message is required.",
      });
    }

    const result = await chatService.handleChatMessage({
      message: message.trim(),
      userId,
      conversationId,
    });

    return res.status(200).json({
      success: true,
      reply: result.reply,
      conversationId: result.conversationId,
    });
  } catch (error) {
    const statusCode = error.statusCode || 500;

    return res.status(statusCode).json({
      success: false,
      message: error.message || "Failed to process chat message.",
    });
  }
};

export const getChatHistory = async (req, res) => {
  try {
    const { conversationId } = req.params;
    const userId = req.user._id; // Securely verify ownership

    if (!isValidString(conversationId)) {
      return res.status(400).json({
        success: false,
        message: "Conversation ID is required.",
      });
    }

    const history = await chatService.getConversationHistory(conversationId, userId);

    return res.status(200).json({
      success: true,
      data: history,
    });
  } catch (error) {
    return res.status(500).json({
      success: false,
      message: error.message || "Failed to fetch chat history.",
    });
  }
};

File: AI/AI/app/controllers/chat_controller.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import Optional
from bson import ObjectId

import app.services.chat_service as chat_service

# TODO: ntar migrasi ke mongodb kalo be udh siap
fake_db = []

async def get_db():
    yield fake_db

class ChatRequest(BaseModel):
    message: str
    user_id: Optional[str] = None
    conversation_id: Optional[str] = None

async def send_chat_message(request: ChatRequest, db=Depends(get_db)):
    if not request.message or not request.message.strip():
        raise HTTPException(status_code=400, detail="Message is required.")
        
    try:
        result = await chat_service.handle_chat_message(
            db=db,
            message=request.message.strip(),
            user_id=request.user_id,
            conversation_id=request.conversation_id
        )
        return {
            "success": True,
            "reply": result["reply"],
            "conversation_id": result["conversation_id"]
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

async def get_chat_history(conversation_id: str, db=Depends(get_db)):
    if not conversation_id or not conversation_id.strip():
        raise HTTPException(status_code=400, detail="Conversation ID is required.")
        
    try:
        history = await chat_service.get_conversation_history(db, conversation_id)
        
        for msg in history:
            if "_id" in msg:
                msg["_id"] = str(msg["_id"])
                
        return {
            "success": True,
            "data": history
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
